### ISSUE TITLE
Add option to install only dev dependencies

### ISSUE THREAD

**Issue by jdinunzio** [python-poetry/poetry#2572]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Feature Request

Add an option to `poetry install` to install only dev dependencies, e.g. `poetry install --dev-only`.

Use Case:

Being able to quickly install only dev dependencies in a Continuous Integration environment to run basic quality checks as fast as possible.

As a developer, I run a series of quick checks on my code, like pylint, flake, black, and mypy, both locally and in a CI environment. Locally, those tools can be installed by running `poetry install`  together with the rest of the package's dependencies. 

In a CI environment I would like to run those tests as quickly as possible to minimize the reponse time of the CI pipeline in the detection of problem in the code. Making a full install of all dependencies could be time consuming for certain projects.

Adding an option to install dev dependencies only will simplify the scripts that run those first stages in the CI pipeline.

**Comment by irfanahm3d** [comment]
Likewise. Facing a similar scenario and I feel it would be a great feature to have! 🙂

**Comment by dannywillems** [comment]
Same here

**Comment by nniehoff** [comment]
Even though PR #2917 has been merged the option still does not exist in 1.1.6.

**Comment by finswimmer** [comment]
Hey @nniehoff,

everything that is in master will be available in poetry 1.2. poetry 1.1 will not receive any new feature.

fin swimmer

**Comment by jose-lpa** [comment]
I'm not sure this has been implemented as people expected it.

I understood from all related issues and PRs that people wanted the same as I actually want: be able to install the development dependencies, **but without altering the main dependencies status**.

What this actually does, as far as I can see, is to make sure that **only the development dependencies are installed**. So it actually **removes any other dependency not specified as `dev`**.

Quick check screenshot:

![Screenshot from 2021-06-29 16-16-18](https://user-images.githubusercontent.com/2226098/123823737-5fc0f380-d8f5-11eb-902d-d9ab2a7fecc9.png)

So as you can see the command `poetry install --dev-only` actually left the project unusable, because it removed all core packages just by not being specified as `dev` dependencies.

I'm not sure what people wanted, but I seriously doubt anyone would expect this behavior when issuing a command to "install development dependencies". 

In my case, what I wanted is to use `poetry install --no-dev` in a Dockerfile to build a container image that can be used in both development, CI and Production. Then use `poetry install --dev-only` in the CI scripts as a previous step before running tests.

**Comment by opelr** [comment]
Hey @jose-lpa, thanks for bringing this up!

Just for my understanding, are you hoping that _all_ project dependencies would get installed in your workflow? I see `--no-dev` and `--dev-only` as being mutually exclusive -- each option specifying both the presence of one set of dependencies with the absence of another -- and, importantly, _not_ being additive, as it seems you desire.

For further clarification, running `poetry install` _does_ install development dependencies. This issue, and others, were requesting a way to _only_ install development dependencies, whereas in your example you're asking for both production and development dependencies.

The current behavior for `--dev-only` lets me run CI scripts which don't depend on my application code, e.g. formatting and static analysis. For your project, I'm wondering if simply running `poetry install` before your CI scripts (since your CI seems coupled to your application code & its dependencies) would solve the problem.

Finally, `--dev-only` is implemented in the same way that `--no-dev` is, in that first running `poetry install` followed by `poetry install --no-dev` removes a subset of installed dependencies. That said, I will defer to the core maintainers to correct me if I'm wrong!

Pinging @finswimmer as they reviewed my previous PR

**Comment by finswimmer** [comment]
You are absolutely right @opelr :+1: Couldn't explain it better :)

**Comment by jdinunzio** [comment]
Hi, opelr!

Sorry for any confusion! Yes, I would like to have a `--dev-only` that install only dev dependencies. Yes, `--dev-only` and `--no-dev` are mutually exclusive, in fact the set of installed packages for those cases is also mutually exclusive. And yes, the use case is to simplify and make faster CI scripts. The only reason I mentioned the default `poetry install` was to stress the fact that the use case is needed in CI environments, and not in dev machines where installation time is not an issue.

I was hoping `--no-dev` could be implemented without a full `poetry install`, since that would defeat the purpose of having only dev tools installed _as soon as possible_ in CI pipelines. 

Thanks for everything!

**Comment by jose-lpa** [comment]
Many thanks for your response @opelr, very appreciated.

## TL;DR
What I want to do is something like this:
`pip install -r production_requirements.txt`
...and in some other place later, in the CI flow...
`pip install -r dev_requirements.txt`

## Why do you want this?

Maybe I don't understand the flow correctly, but it doesn't make sense to me to trigger a command that "installs dev dependencies" and the command actually **uninstalls existing packages that are not in dev**. My intention is to just install a set of packages defined as "development", such as test suites, linting tools, etc. That doesn't necessarily means I want to get rid of core packages when doing so.

I mean, if I have for example a Django project and I want to install at some point Pylint to perform some code checks, Pylint is clearly a non-production package, so I want to get it separated as "development dependency". However, if when I trigger Pylint install with Poetry **Django gets removed as a consequence**, the project remains unusable because one of the core packages is now missing and Pylint (and also any other tool or test suite) is not going to work at all, raising lots of exceptions for `ModuleNotFoundError`, because the core package Django I use all the code I want to lint/test is gone.

On the other hand, I could add Pylint, Pytest and others to "production packages" to use them with the core packages, which effectively makes the "development packages" section useless :shrug:

So, that's why I wanted some option that allows me to install **both** production and dev packages at some point without also installing again the production ones, which would already be installed. This is exactly what Pipenv does, for example, with `pipenv install --dev`.

## But WHY do you want this???

I want it because I have a CI flow that goes like this:

1. Build a Docker image for a web service. Image contains **only** the production needed packages, to make it smaller of course. No need to carry Pylint, Pytest, Black and friends in the container (they are never used when running the service).
2. Run a container for that image.
3. :rotating_light: `docker exec ci_container pip install -r requirements/development.txt` :point_left: this only installs the dev packages **in the running container**, image remains clean with only the core packages. It also doesn't touch the previously installed packages at all. 
4. `docker exec ci_container pylint pytest black ...` all the checks.
5. All good? Then upload **the image** to the container registry.

So, I basically wanted to substitute the `pip` steps by `poetry`, being this tool a bit more advanced in terms of managing interdependencies and other goodies.

Thanks again for the help.

**Comment by opelr** [comment]
@jose-lpa 

Running `poetry install` installs both production and development dependencies -- can you run that in step 3 of your example, or does it not meet your needs? If not, what is the behavior you're seeing?

| Command | Prod. Deps | Dev. Deps |
|---|---|---|
| `poetry install --no-dev` | True | False |
| `poetry install --dev-only` | False | True |
| `poetry install` | **True** | **True** |

**Comment by vchrombie** [comment]
Hi

I understand the use-case of the issue and thanks to @jose-lpa for working on this feature. Looking forward to the 1.2.0 release for using such great features.

Also, if anyone is facing any problem with installing only the dev-dependencies and looking for a workaround, please checkout this tool [vchrombie/peodd](https://github.com/vchrombie/peodd). It can be used to export only the dev-dependencies to a `requirements.txt` format file.

You can install this using pip and export the dev-dependencies to a txt file and later use it to install the dev-dependencies.
```
$ pip install peodd
$ peodd -o requirements-dev.txt
$ pip install requirements-dev.txt
```

Thanks.

**Comment by brandon-leapyear** [comment]
> @jose-lpa
> 
> Running `poetry install` installs both production and development dependencies -- can you run that in step 3 of your example, or does it not meet your needs? If not, what is the behavior you're seeing?
> 
> Command	Prod. Deps	Dev. Deps
> `poetry install --no-dev`	True	False
> `poetry install --dev-only`	False	True
> `poetry install`	**True**	**True**

This does seem to match solve what @jose-lpa was talking about, but there are other scenarios where the current status quo does not work. In our workflow, we would like to have two scripts, one that installs dev deps and one that installs all deps, so that the CI job that does linting can install just dev deps and the CI job that does the build installs everything. Conceptually, these two scripts (installing lint deps and installing build deps) should be commutative and independent. If one builds and then lints, it shouldnt uninstall anything; the only thing the linter cares about is the _presence_ of the linting deps; it couldn't care less if other deps are installed as well.

And more importantly, the two scripts together should be idempotent; if we had a wrapper script that did `lint.sh && build.sh`, running that script repeatedly today would work fine the first time, but then every subsequent time would uninstall and reinstall core deps.

**Comment by dapicester** [comment]
My workaround for this:
```sh
poetry export --without-hashes -o requirements.txt
poetry export --without-hashes --dev -o requirements.all.txt
comm -3 <(sort requirements.txt) <(sort requirements.all.txt) > requirements.dev.txt
```
I generate the two requirements files, then I use the [comm](https://linux.die.net/man/1/comm) command to get the differences.

**Comment by akharrou** [comment]
what about `poetry install DOMAIN`, where domain refers to some `tool.poetry.[dependencies|dev-dependencies|extras]` or other arbitrary `pyproject.toml` section under which a user specifies a list of packages ?

**Comment by dapicester** [comment]
@akharrou is this supported yet?
For me this happens
```bash
$ poetry --version
Poetry version 1.1.14
$ poetry install dependencies
Too many arguments.
```

**Comment by akharrou** [comment]
> @akharrou is this supported yet? For me this happens
> 
> ```shell
> $ poetry --version
> Poetry version 1.1.14
> $ poetry install dependencies
> Too many arguments.
> ```

no sadly it isn't, this was just a suggestion.

**Comment by dapicester** [comment]
Nice! I see more details [here](https://python-poetry.org/docs/master/managing-dependencies/#dependency-groups).

Moreover:
> Poetry will slowly transition away from the dev-dependencies notation which will soon be deprecated, so it’s advised to migrate your existing development dependencies to the new group notation.

And the new CLI will allow to install groups:

    $ poetry install --with docs
 
I am so looking forward to this.
